File: android/app/src/main/kotlin/com/hanjukukobo/walking_analysis/VisualizationUtils.kt
package com.hanjukukobo.walking_analysis

import android.graphics.*

object VisualizationUtils {
    /** Radius of circle used to draw keypoints.  */
    private const val CIRCLE_RADIUS = 6f

    /** Width of line used to connected two keypoints.  */
    private const val LINE_WIDTH = 4f

    /** Pair of keypoints to draw lines between.  */
    private val bodyJoints = listOf(
        Pair(BodyPart.NOSE, BodyPart.LEFT_EYE),
        Pair(BodyPart.NOSE, BodyPart.RIGHT_EYE),
        Pair(BodyPart.LEFT_EYE, BodyPart.LEFT_EAR),
        Pair(BodyPart.RIGHT_EYE, BodyPart.RIGHT_EAR),
        Pair(BodyPart.NOSE, BodyPart.LEFT_SHOULDER),
        Pair(BodyPart.NOSE, BodyPart.RIGHT_SHOULDER),
        Pair(BodyPart.LEFT_SHOULDER, BodyPart.LEFT_ELBOW),
        Pair(BodyPart.LEFT_ELBOW, BodyPart.LEFT_WRIST),
        Pair(BodyPart.RIGHT_SHOULDER, BodyPart.RIGHT_ELBOW),
        Pair(BodyPart.RIGHT_ELBOW, BodyPart.RIGHT_WRIST),
        Pair(BodyPart.LEFT_SHOULDER, BodyPart.RIGHT_SHOULDER),
        Pair(BodyPart.LEFT_SHOULDER, BodyPart.LEFT_HIP),
        Pair(BodyPart.RIGHT_SHOULDER, BodyPart.RIGHT_HIP),
        Pair(BodyPart.LEFT_HIP, BodyPart.RIGHT_HIP),
        Pair(BodyPart.LEFT_HIP, BodyPart.LEFT_KNEE),
        Pair(BodyPart.LEFT_KNEE, BodyPart.LEFT_ANKLE),
        Pair(BodyPart.RIGHT_HIP, BodyPart.RIGHT_KNEE),
        Pair(BodyPart.RIGHT_KNEE, BodyPart.RIGHT_ANKLE)
    )

    // Draw line and point indicate body pose
    fun drawBodyKeyPoints(
        input: Bitmap,
        persons: List<Person>
    ): Bitmap {
        val paintCircle = Paint().apply {
            strokeWidth = CIRCLE_RADIUS
            color = Color.RED
            style = Paint.Style.FILL
        }
        val paintLine = Paint().apply {
            strokeWidth = LINE_WIDTH
            color = Color.RED
            style = Paint.Style.STROKE
        }

        val output = input.copy(Bitmap.Config.ARGB_8888, true)
        val originalSizeCanvas = Canvas(output)
        persons.forEach { person ->
            bodyJoints.forEach {
                val pointA = person.keyPoints[it.first.position].coordinate
                val pointB = person.keyPoints[it.second.position].coordinate
                originalSizeCanvas.drawLine(pointA.x, pointA.y, pointB.x, pointB.y, paintLine)
            }

            person.keyPoints.forEach { point ->
                originalSizeCanvas.drawCircle(
                    point.coordinate.x,
                    point.coordinate.y,
                    CIRCLE_RADIUS,
                    paintCircle
                )
            }

            var pointF = arrayOf<PointF>()
            for (i in 0 .. 5 step 1) {
                pointF += person.keyPoints[i+11].coordinate
            }
        }

        return output
    }
}

File: ios/Runner/VisualizationUtils.swift
//
//  Visualization.swift
//  Runner
//
//  Created by hanzyuku on 2022/08/02.
//

import Foundation

// シングルトンクラスとして実装
final public class Visualizationutils {
    private init() {}
    public static let shared = Visualizationutils()
    
    private let CIRCLE_RADIUS = 6
    private let LINE_WIDTH = 4
    
    private let bodyJoints = [
        (BodyPart.NOSE, BodyPart.LEFT_EYE),
        (BodyPart.NOSE, BodyPart.RIGHT_EYE),
        (BodyPart.LEFT_EYE, BodyPart.LEFT_EAR),
        (BodyPart.RIGHT_EYE, BodyPart.RIGHT_EAR),
        (BodyPart.NOSE, BodyPart.LEFT_SHOULDER),
        (BodyPart.NOSE, BodyPart.RIGHT_SHOULDER),
        (BodyPart.LEFT_SHOULDER, BodyPart.LEFT_ELBOW),
        (BodyPart.LEFT_ELBOW, BodyPart.LEFT_WRIST),
        (BodyPart.RIGHT_SHOULDER, BodyPart.RIGHT_ELBOW),
        (BodyPart.RIGHT_ELBOW, BodyPart.RIGHT_WRIST),
        (BodyPart.LEFT_SHOULDER, BodyPart.RIGHT_SHOULDER),
        (BodyPart.LEFT_SHOULDER, BodyPart.LEFT_HIP),
        (BodyPart.RIGHT_SHOULDER, BodyPart.RIGHT_HIP),
        (BodyPart.LEFT_HIP, BodyPart.RIGHT_HIP),
        (BodyPart.LEFT_HIP, BodyPart.LEFT_KNEE),
        (BodyPart.LEFT_KNEE, BodyPart.LEFT_ANKLE),
        (BodyPart.RIGHT_HIP, BodyPart.RIGHT_KNEE),
        (BodyPart.RIGHT_KNEE, BodyPart.RIGHT_ANKLE)
    ]
    
    func drawBodyKeyPoints(cgContext: CGContext, inputImage: CGImage, persons: [Person]) -> CGImage? {
        // 線を描画するためのセットアップ
        cgContext.setLineWidth(CGFloat(LINE_WIDTH))
        cgContext.setStrokeColor(UIColor.red.cgColor)
        
        // 丸を描画するためのセットアップ
        cgContext.setFillColor(UIColor.red.cgColor)
        
        persons.forEach {
            let person = $0
            
            bodyJoints.forEach {
                let pointA = person.keyPoints[$0.0.rawValue].coordinate
                let pointB = person.keyPoints[$0.1.rawValue].coordinate
                cgContext.move(to: CGPoint(x: pointA.x, y: pointA.y))
                cgContext.addLine(to: CGPoint(x: pointB.x, y: pointB.y))
            }
            
            person.keyPoints.forEach {
                let rectangle = CGRect(
                    origin: $0.coordinate,
                    size: CGSize(width: CIRCLE_RADIUS, height: CIRCLE_RADIUS))
                cgContext.addEllipse(in: rectangle)
            }
        }
        
        guard let image = cgContext.makeImage() else { return nil }
        return image
    }
}
